packages/block-editor/src/store/private-selectors.js

```diff
@@ -114,6 +114,7 @@ export const getEnabledClientIdsTree = createSelector(
 		state.blockEditingModes,
 		state.settings.templateLock,
 		state.blockListSettings,
+		state.editorMode,
```

Comment:
List View depends on this selector to determine which clientIds to display. As it's memoized, when we switch editor modes we will need to recompute the value to ensure it's accurate.

Reply:
I'd really like a confidence check on this one. I would have thought that having the `state.blockEditingModes` dependencies would be enough here but apparently not 😕

Reply:
@talldan This turns out to still be required otherwise List View doesn't not reflect the changes correctly. Going to have to continue digging to work out why though...

Reply:
To define the dependencies of a selector, we need to just look what selectors it uses and what state it users So we need to look at. `getEnabledClientIdsTreeUnmemoized`. So basically we need this to be the sum of `getBlockEditingMode`'s dependencies and `getBlockOrder`'s dependencies.

so `state.blocks.order` + `state.editorMode` + `state.settings?.[ sectionRootClientIdKey ]` + `state.settings.templateLock` + `state.blockEditingModes` + `state.blocks.parents` (same thing as order though, just denormalized) + `state.blockListSettings` + `state.blocks.byClientId` + `state.blocks.attributes` (for native only) + registered blocks.

It's a very hard selector to memoize IMO. We have options:

 - Keep as is and risk re-renderings that are not needed that often.
 - Remove the need for this selector (probably involves component refactoring)
 - Denormalize the selector in a dedicated redux state (like we do for blocks.tree...) it's very hard to get right though and I would only do it if we realize this selector is a real bottleneck performance wise.

Reply:
Yeah, this is missing a lot of dependencies, not just `state.editorMode`. There's a good chance that there's other subtle bugs because of this, but adding this many dependencies pretty much negates the usefulness of the memoization.

For now, I think we should keep things as-is to fix the obvious bug of not updating the list view correctly. The `editorMode` doesn't get updated that frequently, so I don't think it will have that large of a performance impact.

Reply:
> To define the dependencies of a selector, we need to just look what selectors it uses

I came to the same conclusion about dependencies here.

https://github.com/WordPress/gutenberg/pull/65202#discussion_r1754704128

> It's a very hard selector to memoize IMO 

Agreed.

> Keep as is and risk re-renderings that are not needed that often.
> Remove the need for this selector (probably involves component refactoring)
> Denormalize the selector in a dedicated redux state (like we do for blocks.tree...) it's very hard to get right though and I would only do it if we realize this selector is a real bottleneck performance wise.

`editorMode` doesn't change often and I can't imagine that changing in the future. We're ~never~ very unlikely going to change `editorMode` on an action that runs repeatedly as it would be poor for UX.

On balance I'd say running with the fix in this PR would be an acceptable trade off.

Would it help if I added a comment to the selector to note that it may be missing dependencies and referring to this discussion? That way if it becomes a problem in the future we know where to look.

Reply:
I think it's definitely acceptable to run with this PR, my worry is that one of the underlying selectors change and we don't update the dependencies here. It's indirect.

Reply:
I agree.

The only viable options seems to be:

> Remove the need for this selector (probably involves component refactoring)

It would be cool if you could declare a selector as dependency and then it would automatically use that selector's dependencies as dependencies.

Reply:
I think you can kind of do that:
https://github.com/WordPress/gutenberg/blob/9bdfebfa6760fcc9f1d11a185ecb2940d475fcb5/packages/block-editor/src/store/selectors.js#L850

But I'm not sure it'd work here because of the way it depends on `blockEditingMode` (unmemoized, and also per block), which then depends on `editorMode`. 🤔